sympy/matrices/sparse.py

```diff
@@ -479,8 +481,25 @@ def _eval_matrix_mul(self, other, **kwargs):
                 # these are the only things that need to be multiplied.
                 indices = set(col_lookup[col].keys()) & set(row_lookup[row].keys())
                 if indices:
-                    val = sum(row_lookup[row][k]*col_lookup[col][k] for k in indices)
-                    smat[row, col] = val
+                    len_indices = len(indices)
+                    vec = [None]*len_indices
+                    for i, k in enumerate(indices):
+                        c       = row_lookup[row][k]*col_lookup[col][k]
+                        _expand = mulsimp and getattr(c, 'expand', None)
+                        vec[i]  = _expand(power_base=False, power_exp=False, log=False, \
+                                multinomial=False, basic=False) if _expand else c
```

Comment:
The implementation of `_eval_matrix_mul` here for sparse matrices seems to duplicate a lot of the code used for dense matrices. Can we not factor this out?

Reply:
It is actually different in how it goes about its job and it seems like it would be much faster for large sparse matrices, again I added it here so if someone multiplies a Sparse with a Dense it will still simplify.

Reply:
Do you mean if someone does `Msparse.mul(Mdense, mulsimp=True)`?

Reply:
Yes, if it is not here then a Sparse.mul(Dense, mulsimp=True) will not simplify because it is done in this routine. If you want I will remove it from MatrixArithmetic._eval_matrix_mul, but here it is important.

Reply:
I don't suggest to remove it here but seeing that there is some duplication of code it would be good that could be separated so that the code is in one place somewhere. The problem with having the same code twice is that it will end up out of sync. I think that this is generally a problem for sparse because it is less used and possibly less tested.

Reply:
I thought about that but since it is such a small chunk of code I didn't want to add overhead and I couldn't separate the parent routines because they do actually work differently. For example MatrixArithmetic must access the matrix elements by individual indexing and not an array like Dense, and Sparse actually works with a dictionary of positions of elements.

Reply:
That is to say couldn't join the parent routines...

Reply:
If I understand correctly essentially what you have defined here is a sum method that applies some simplification to each term and then to the sum as a whole. It could be abstracted out to a function that behaves like `sum(things)` but does this simplification. A `sumsimp(things)` function might be a more appropriate addition than `fastalgsimp` since `fastalgsimp` is really designed to be used in conjuction with the expand+sum here.

Reply:
I could change it to use pure indexing which would abstract MatrixArithmetic, slow Dense a little, but it would totally throw away the advantage of using Sparse over Dense. Let me see if there is a simple way to can centralize it a little.

Reply:
I was thinking something like:
```python
def _eval_matrix_mul(self, other, mulsimp=None):

    def entry(i, j):
        terms = (self[i,k]*other[k,j] for k in range(self.cols))
        return sumsimp(terms, simplify=mulsimp)
    
    return self._new(self.rows, other.cols, entry)

    
def sumsimp(terms, simplify=True):
    
    def expand(term):
        _expand = getattr(term, 'expand', None)
        if _expand is not None:
            term = _expand(power_base=False, power_exp=False, log=False,
                    multinomial=False, basic=False) 
        return term

    if simplify:
        terms = map(expand, terms)

    total = reduce(lambda a, b: a+b, terms)

    if simplify:
        total = fastalgsimp(total)

    return total
```

Reply:
Again, this would throw out the point of using Sparse matrices, but I am on to something to pull out the inner loop into one place.

Reply:
Pulled out the common inner portion leaving the code unique to each matrix mul routine in place.

Reply:
That was intended to be the dense version. The sparse one just has a different way of defining the terms so that the loops there become:
```python
    smat = {}
    for row in row_lookup.keys():
        for col in col_lookup.keys():
            # find the common indices of non-zero entries.
            # these are the only things that need to be multiplied.
            indices = set(col_lookup[col].keys()) & set(row_lookup[row].keys())
            if indices:
                terms = (row_lookup[row][k]*col_lookup[col][k] for k in indices)
                smat[row, col] = sumsimp(terms, mulsimp)
```

Reply:
In any case I pulled out the important bit so that functionality is centralized and the same inner loop is used by all three _eval_matrix_mul.

Reply:
The version you have just pushed is very similar to what I proposed above except that it has a different signature which requires passing in a vec list. Since it is easily possible (see above) to avoid that we can make this a more reusable function.

Also this function can have a more generic use than just for matrices. There can be many other places where things are added and we want to control expression blowup. I actually think that this function you have extracted is *the* important simplification routine from this PR so I would suggest making fastalgsimp a private/internal routine for use in this one.

Reply:
I pass the vec list to avoid instantiating a new list for each element since I know the size ahead of time, it could save some overhead but I can also make that optional to make this more generic. I have nothing against internalizing fastalgsimp and turning this into a more general sum-of-products routine. In which case would you want me to move it into simplify?

Reply:
Look at the code I've shown above: no list is instantiated since none is needed (although to be fair I haven't tested that code).

You're probably right that it belongs in simplify.

What particularly interests me about this right now is that I can see this routine being directly usable for blowup control in other places like rref, maybe det etc. so I think it would be good to get the signature right for reuse.

Reply:
When you say you worry about the signature is that for caching or for some other reason?

Reply:
I just think that a function shouldn't have unnecessary parameters that make it awkward to use consider the difference between:
```python
terms = ...
total = sumsimp(terms, simplify=True)

terms = ...
terms = list(terms) # convert e.g. generator to a list for len
vec = [None] * len(terms)
total = matrix_inner_loop(vec, terms, mulsimp=True)
```
To me it is clear that the former is easier to use, understand, document etc. A user would rightly complain about having to create the vec list when it isn't actually needed and it can be created internally using the information provided by the other arguments.

An API shouldn't require redundant information from the caller unless there is a good reason.

As in aside there is rarely any significant performance benefit for preallocating lists in Python. The list data structure is designed to grow efficiently so that idiomatic use is append or a list comprehension. I very much doubt that any performance improvement this could give would be measurable in the context of the other operations here.

Reply:
I am not particularly tied to that vec it simply came over from previous implementation of the _eval_matrix_mul routines, already thinking of doing it differently and even cleaner so no worries, will send it up tomorrow probably.